aten/src/ATen/Quantizer.cpp

```diff
@@ -0,0 +1,134 @@
+    allocator,
+    /*resizable=*/true);
+  auto quantizer = make_per_layer_affine_quantizer(scale, zero_point);
+  auto tensor = detail::make_tensor<QTensorImpl>(
+      storage_impl, at::CPUTensorId(), false, quantizer);
+  // Default TensorImpl has size [0]
+  if (sizes.size() != 1 || sizes[0] != 0) {
+    get_qtensorimpl(tensor)->set_sizes_contiguous(sizes);
+  }
+  return tensor;
+}
+
+void ChooseParams(RealTensor tensor, float* r_scale, int* r_zero_point) {
+  // TODO: dispatch according to tensor.dtype()
+
+  RealTensor real_max_t_ = tensor.max();
+  RealTensor real_min_t_ = tensor.min();
+  float real_max = real_max_t_.data<float>()[0];
+  float real_min = real_min_t_.data<float>()[0];
+  // We extend the [min, max] interval to ensure that it contains 0.
+  // Otherwise, we would not meet the requirement that 0 be an exactly
+  // representable value.
+  auto min = std::min(real_min, 0.f);
+  auto max = std::max(real_max, 0.f);
+
+  // the min and max quantized values, as floating-point values
+  const float qmin = 0;
+  const float qmax = 255;
+
+  // First determine the scale.
+  const double scale = (max - min) / (qmax - qmin);
+
+  // Zero-point computation.
+  // First the initial floating-point computation. The zero-point can be
+  // determined from solving an affine equation for any known pair
+  // (real value, corresponding quantized value).
+  // We know two such pairs: (rmin, qmin) and (rmax, qmax).
+  // Let's use the first one here.
+  const double initial_zero_point = qmin - min / scale;
+
+  // Now we need to nudge the zero point to be an integer
+  // (our zero points are integer, and this is motivated by the requirement
+  // to be able to represent the real value "0" exactly as a quantized value,
+  // which is required in multiple places, for example in Im2col with SAME
+  // padding).
+  std::uint8_t nudged_zero_point = 0;
+  if (initial_zero_point < qmin) {
+    nudged_zero_point = qmin;
+  } else if (initial_zero_point > qmax) {
+    nudged_zero_point = qmax;
+  } else {
+    nudged_zero_point =
+      static_cast<std::uint8_t>(std::round(initial_zero_point));
+  }
+  *r_scale = scale;
+  *r_zero_point = int(nudged_zero_point);
+}
+
+qint8 QuantizeUint8(float scale, int32_t zero_point, float value) {
+  const int32_t qmin = std::numeric_limits<uint8_t>::min();
+  const int32_t qmax = std::numeric_limits<uint8_t>::max();
+
+  auto r = zero_point + static_cast<int32_t>(Round(value / scale));
+  r = std::max(r, qmin);
+  r = std::min(r, qmax);
+  return static_cast<qint8>(r);
+}
+
+QTensor PerLayerAffineQuantizer::quantize(RealTensor tensor) {
+  //ChooseParams(tensor, &scale_, &zero_point_);
+
+  IntArrayRef sizes = tensor.sizes();
+  QTensor qv = new_qtensor(sizes, tensor.options().dtype(at::kQInt8), scale_, zero_point_);
+  auto qvd = qv.data<qint8>();
+  const float* svd = tensor.data<float>();
+  for (int i = 0; i < tensor.numel(); ++i) {
+    qvd[i] = QuantizeUint8(scale_, zero_point_, svd[i]);
```

Comment:
Companion to previous comment: Pass reciprocal of scale to avoid repetitive divides in for loop.